### PR TITLE
Fix deprecated SecurityContext usage on loginManager

### PR DESCRIPTION
**Pull request by Soullivaneuh**
Fix deprecated `SecurityContext` usage on loginManager.

Symfony 2.3+ BC kept.

@xabbuh: I search a long time on PR, nobody already did this normally. :-)

### CODE REVIEW

File: Resources/config/security.xml [hunk 1]
```diff
@@ -6,7 +6,6 @@
 
     <parameters>
         <parameter key="fos_user.security.interactive_login_listener.class">FOS\UserBundle\EventListener\LastLoginListener</parameter>
-        <parameter key="fos_user.security.login_manager.class">FOS\UserBundle\Security\LoginManager</parameter>
```
Comment (xabbuh) [inline Resources/config/security.xml]: removing the parameter is a BC break

File: DependencyInjection/Compiler/LoginManagerPass.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        // Set the SecurityContext for Symfony <2.6
+        // Should go back to simple xml configuration after <2.6 support
+        $tokenStorageReference = interface_exists('Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface')
+            ? new Reference('security.token_storage') : new Reference('security.context');
```
Comment (xabbuh) [inline DependencyInjection/Compiler/LoginManagerPass.php]: You don't need a compiler pass for this, but you can do this in the `FOSUserExtension` instead.

File: Resources/config/security.xml (hunk 1 above)
Comment (Soullivaneuh) [inline Resources/config/security.xml]: I'm curious. Why this is BC break?

I tried to follow [this advice](https://github.com/FriendsOfSymfony/FOSOAuthServerBundle/pull/315#discussion_r29235933).

What do you think @stof?

Comment (stof) [inline Resources/config/security.xml]: @Soullivaneuh please don't remove existing parameters. This is a BC break, so I would do it all at once, not service by service

File: Security/LoginManager.php [hunk 3]
```diff
@@ -27,16 +28,19 @@
...
+    private $tokenStorage;
     private $userChecker;
     private $sessionStrategy;
     private $container;
 
-    public function __construct(SecurityContextInterface $context, UserCheckerInterface $userChecker,
+    public function __construct($tokenStorage, UserCheckerInterface $userChecker,
                                 SessionAuthenticationStrategyInterface $sessionStrategy,
                                 ContainerInterface $container)
     {
```
Comment (stof) [inline Security/LoginManager.php]: you should add a check in the controller to accept only these 2 interfaces (see how Symfony 2.6 does it)

File: Resources/config/security.xml (hunk 1 above)
Comment (stof) [inline Resources/config/security.xml]: And my advice was talking about *new* parameters

File: DependencyInjection/Compiler/LoginManagerPass.php (hunk 2 above)
Comment (Soullivaneuh) [inline DependencyInjection/Compiler/LoginManagerPass.php]: Done.

File: Resources/config/security.xml (hunk 1 above)
Comment (Soullivaneuh) [inline Resources/config/security.xml]: Ok so I will revert my [changes](https://github.com/FriendsOfSymfony/FOSOAuthServerBundle/pull/315/files#diff-4d3f32e7db124020203dccf0e273e276L7) on this another PR too.

Comment (Soullivaneuh) [inline Resources/config/security.xml]: But why is this BC break? Because some user could use those parameters?

File: Security/LoginManager.php (hunk 3 above)
Comment (Soullivaneuh) [inline Security/LoginManager.php]: Have you a code sample reference please? Thanks.

Comment (stof) [inline Security/LoginManager.php]: I don't find the place where we used it (components have bumped their interdependencies, and I cannot find the bundle where we used it). Here is what I'm asking for:

```php
if (!$tokenStorage instanceof TokenStorageInterface && !$tokenStorage instanceof SecurityContextInterface) {
    throw new \InvalidArgumentException('...');
}
```

Comment (Soullivaneuh) [inline Security/LoginManager.php]: Done. :+1:

**Comment by Soullivaneuh** [comment]
Corrections done, ready for merge.

File: Security/LoginManager.php [hunk 4]
```diff
@@ -27,16 +28,23 @@
...
-        $this->securityContext = $context;
+        if (!$tokenStorage instanceof TokenStorageInterface && !$tokenStorage instanceof SecurityContextInterface) {
+            throw new \InvalidArgumentException(sprintf('Argument 1 of %s should be an instance of TokenStorageInterface or SecurityContextInterface', __CLASS__));
```
Comment (stof) [inline Security/LoginManager.php]: FQCN should be used (and IMO, there is no need to mention ``__CLASS__``. the error location already shows you where the error is

Comment (Soullivaneuh) [inline Security/LoginManager.php]: Done.

File: DependencyInjection/FOSUserExtension.php [hunk 5]
```diff
@@ -55,6 +55,15 @@ public function load(array $configs, ContainerBuilder $container)
...
         }
 
+        // Set the SecurityContext for Symfony <2.6
+        // Should go back to simple xml configuration after <2.6 support
+        $tokenStorageReference = interface_exists('Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface')
+            ? new Reference('security.token_storage') : new Reference('security.context');
+        $container
+            ->getDefinition('fos_user.security.login_manager')
+            ->replaceArgument(0, $tokenStorageReference)
+        ;
```
Comment (xabbuh) [inline DependencyInjection/FOSUserExtension.php]: Imho this would be more readable if you didn't use the ternary operator.

File: FOSUserBundle.php [hunk 6]
```diff
@@ -11,13 +11,14 @@
...
-use FOS\UserBundle\DependencyInjection\Compiler\RegisterMappingsPass;
+use Doctrine\Bundle\CouchDBBundle\DependencyInjection\Compiler\DoctrineCouchDBMappingsPass;
 use Doctrine\Bundle\DoctrineBundle\DependencyInjection\Compiler\DoctrineOrmMappingsPass;
 use Doctrine\Bundle\MongoDBBundle\DependencyInjection\Compiler\DoctrineMongoDBMappingsPass;
-use Doctrine\Bundle\CouchDBBundle\DependencyInjection\Compiler\DoctrineCouchDBMappingsPass;
+use FOS\UserBundle\DependencyInjection\Compiler\LoginManagerPass;
+use FOS\UserBundle\DependencyInjection\Compiler\RegisterMappingsPass;
+use FOS\UserBundle\DependencyInjection\Compiler\ValidationPass;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\HttpKernel\Bundle\Bundle;
```
Comment (xabbuh) [inline FOSUserBundle.php]: You can completely revert the changes here (in fact you must at least remove the `LoginManagerPass`).

Comment (Soullivaneuh) [inline FOSUserBundle.php]: It's an imports optimization done by my IDE. I removed `LoginManagerPass`.

**Comment by Soullivaneuh** [comment]
Code updated following @xabbuh suggestions.

Ping @stof: I think it's ready for merge. :+1:

**Comment by stof** [comment]
Thank you @Soullivaneuh.
